packages/ts/src/components/treemap/index.ts

```diff
@@ -0,0 +1,121 @@
+import { group, max } from 'd3-array'
+import { scaleLinear } from 'd3-scale'
+import { ComponentCore } from 'core/component'
+import { SeriesDataModel } from 'data-models/series'
+import { getColor } from 'utils/color'
+import { getString, getNumber, isNumber } from 'utils/data'
+import { smartTransition } from 'utils/d3'
+import { TreemapConfigInterface, TreemapDefaultConfig } from './config'
+import { TreemapNode } from './types'
+
+import * as s from './style'
+
+export class Treemap<Datum> extends ComponentCore<Datum[], TreemapConfigInterface<Datum>> {
+  static selectors = s
+  protected _defaultConfig = TreemapDefaultConfig as TreemapConfigInterface<Datum>
+  public config: TreemapConfigInterface<Datum> = this._defaultConfig
+
+  datamodel: SeriesDataModel<Datum> = new SeriesDataModel()
+  tiles: Selection<SVGGElement, unknown, SVGGElement, unknown>
+
+  constructor (config?: TreemapConfigInterface<Datum>) {
+    super()
+    if (config) this.setConfig(config)
+    this.tiles = this.g.append('g')
+  }
+
+  _render (customDuration?: number): void {
+    const { config, datamodel: { data } } = this
+    const duration = isNumber(customDuration) ? customDuration : config.duration
+
+    if (!config.layers?.length) {
+      console.warn('Unovis | Treemap: No layers defined')
+      return
+    }
+
+    const layerAccessors = config.layers.map(layerAccessor => (i: number) => getString(data[i], layerAccessor, i))
+    const nestedData = group(data.keys(), ...layerAccessors as [(d: number) => string])
+
+    const rootNode = config.value !== undefined
+      ? hierarchy(nestedData).sum(index => typeof index === 'number' && getNumber(data[index], config.value, index))
+      : hierarchy(nestedData).count()
+
+    const treemapLayout = treemap()
+      .size([this._width, this._height])
+      .round(true)
+      .padding(this.config.padding)
+
+    const treemapData = treemapLayout(rootNode) as TreemapNode<Datum>
+
+    const descendants = treemapData.descendants()
+
+    const opacity = scaleLinear()
+      .domain([1, max(descendants, d => d.depth)])
+      .range([1, 0.2])
+
+    treemapData
+      .eachBefore((node) => {
+        if (!node.children) return
+        node.children.forEach((child, i) => {
+          const treemapChild = child as TreemapNode<Datum>
+          const color = getColor(treemapChild, config.tileColor, i, treemapChild.depth !== 1)
+          treemapChild._state = {
+            fill: color ?? (node as TreemapNode<Datum>)._state?.fill,
+            fillOpacity: color === null ? opacity(treemapChild.depth) : null,
+          }
+        })
+      })
+
+    // Render tiles
+    const visibleNodes = descendants.filter(d => d.depth > 0)
+    const tiles = this.tiles
+      .selectAll<SVGGElement, TreemapNode<Datum>>('g')
+      .data(visibleNodes, d => d._id)
+    const tilesEnter = tiles.enter().append('g')
+
+    // Background rectangles
+    tilesEnter.append('rect').classed('background', true)
```

Comment:
We generally use generated classnames everywhere in Unoivis. This allows users to attach events to them.